File: ios/Reguerta/Reguerta/Data/Products/InMemoryProductRepository.swift
import Foundation

actor InMemoryProductRepository: ProductRepository {
    private var productsById: [String: Product]

    init(items: [Product] = []) {
        self.productsById = Dictionary(uniqueKeysWithValues: items.map { ($0.id, $0) })
    }

    func allProducts() async -> [Product] {
        productsById.values.sorted {
            if $0.archived != $1.archived {
                return !$0.archived && $1.archived
            }
            return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
        }
    }

    func products(vendorId: String) async -> [Product] {
        productsById.values
            .filter { $0.vendorId == vendorId }
            .sorted {
                if $0.archived != $1.archived {
                    return !$0.archived && $1.archived
                }
                return $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending
            }
    }

    func upsert(product: Product) async -> Product {
        productsById[product.id] = product
        return product
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/data/products/InMemoryProductRepository.kt
package com.reguerta.user.data.products

import com.reguerta.user.domain.products.Product
import com.reguerta.user.domain.products.ProductRepository
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class InMemoryProductRepository(
    items: List<Product> = emptyList(),
) : ProductRepository {
    private val mutex = Mutex()
    private val products = items.associateBy { it.id }.toMutableMap()

    override suspend fun getAllProducts(): List<Product> = mutex.withLock {
        products.values.sortedWith(compareBy<Product> { it.archived }.thenBy { it.name.lowercase() })
    }

    override suspend fun getProductsForVendor(vendorId: String): List<Product> = mutex.withLock {
        products.values
            .filter { it.vendorId == vendorId }
            .sortedWith(compareBy<Product> { it.archived }.thenBy { it.name.lowercase() })
    }

    override suspend fun upsertProduct(product: Product): Product = mutex.withLock {
        products[product.id] = product
        product
    }
}
